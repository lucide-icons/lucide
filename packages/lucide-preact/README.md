# Lucide Preact

Implementation of the lucide icon library for preact applications.

> What is lucide? Read it [here](https://github.com/lucide-icons/lucide#what-is-lucide).

## Installation

```sh
yarn add lucide-preact

# or

npm install lucide-preact
```

## How to use

It's build with ESmodules so it's completely threeshakable.
Each icon can be imported as a preact component.

### Example

You can pass additional props to adjust the icon.

``` js
import { Camera } from 'lucide-preact';
// Returns PreactComponent

// Usage
const App = () => {
  return <Camera color="red" size={48}/>
};

export default App;
```

### Props

|  name        |   type   |  default
| ------------ | -------- | --------
| `size`       | *Number* | 24
| `color`      | *String* | currentColor
| `strokeWidth`| *Number* | 2

### Custom props / svg attributes

You can also pass custom props that will be added in the as attributes. With that you can modify the icons look by passing svg attributes.

``` js
// Usage
const App = () => {
  return <Camera fill="red" stroke-linejoin="bevel"/>
};
```

> svg attributes in preact aren't transformed, so if want to change e.g. the `stroke-linejoin` you need to pass it in kebabcase, the way svg spec is written so. See this topic in the [preact documentation](https://preactjs.com/guide/v10/differences-to-react/#svg-inside-jsx).

### One generic icon component

It is possible to create one generic icon component to load icons.

> :warning: Example below importing all EsModules, caution  using this example, not recommended when you using bundlers, your application build size will grow strongly.

#### Icon Component Example

``` js
import * as icons from 'lucide-preact';

const Icon = ({name, color, size}) => {
  const LucideIcon = icons[name];

  return <LucideIcon color={color} size={size} />
};

export default Icon;
```
