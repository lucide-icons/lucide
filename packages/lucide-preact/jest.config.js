module.exports = {
  verbose: true,
  roots: ['<rootDir>/src/', '<rootDir>/tests/'],
  moduleFileExtensions: ['js'],
  transformIgnorePatterns: [`/node_modules`],
  transform: {
    '^.+\\.js$': 'babel-jest',
  },
};
