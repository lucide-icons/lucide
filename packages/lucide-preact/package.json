{
  "name": "lucide-preact",
  "description": "Lucide Preact package, Lucide is a community-run fork of Feather Icons, open for anyone to contribute icons.",
  "version": "0.15.10-beta.7",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide-preact"
  },
  "author": "Eric Fennis",
  "amdName": "lucide-preact",
  "main": "dist/cjs/lucide-preact.js",
  "main:umd": "dist/umd/lucide-preact.js",
  "module": "dist/esm/lucide-preact.js",
  "unpkg": "dist/umd/lucide-preact.min.js",
  "typings": "dist/lucide-preact.d.ts",
  "scripts": {
    "build": "yarn clean && yarn build:icons && yarn build:es && yarn build:types && yarn build:bundles",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.js",
    "build:icons": "yarn --cwd ../../ build:icons --output=../packages/lucide-preact/src --templateSrc=../packages/lucide-preact/scripts/exportTemplate --renderUniqueKey",
    "build:es": "babel src -d dist/esm",
    "build:types": "yarn --cwd ../../ babel-node packages/lucide-preact/scripts/buildTypes.js",
    "build:bundles": "yarn --cwd ../../ rollup -c packages/lucide-preact/rollup.config.js",
    "test": "jest"
  },
  "devDependencies": {
    "@testing-library/preact": "^2.0.1",
    "babel-preset-preact": "^2.0.0",
    "jest": "^26.6.3",
    "preact": "^10.5.13"
  },
  "peerDependencies": {
    "preact": "^10.5.13"
  }
}
