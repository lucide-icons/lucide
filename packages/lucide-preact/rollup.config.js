import plugins from '../../rollup.plugins';
import pkg from './package.json';

const packageName = 'LucidePreact';
const outputFileName = 'lucide-preact';
const rootDir = 'packages/lucide-preact'; // It runs from the root
const outputDir = `${rootDir}/dist`;
const inputs = [`${rootDir}/src/lucide-preact.js`];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify }) =>
    inputs.map(input => ({
      input,
      plugins: plugins(pkg, minify),
      external: ['preact', 'prop-types', 'lucide'],
      output: {
        name: packageName,
        file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
        format,
        sourcemap: true,
        globals: {
          preact: 'preact',
          'prop-types': 'PropTypes',
        },
      },
    })),
  )
  .flat();

export default configs;
