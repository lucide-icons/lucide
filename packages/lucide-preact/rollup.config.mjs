import plugins, { replace } from '@lucide/rollup-plugins';
import pkg from './package.json' assert { type: "json" };

const packageName = 'LucidePreact';
const outputFileName = 'lucide-preact';
const outputDir = 'dist';
const inputs = [`src/lucide-preact.ts`];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
    aliasesSupport: true
  },
  {
    format: 'esm',
    inputs,
    outputDir,
    preserveModules: true,
    aliasesSupport: true
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify, preserveModules, aliasesSupport }) =>
    inputs.map(input => ({
      input,
      plugins: [
        ...(
          !aliasesSupport ? [
            replace({
              "export * from './aliases';": '',
              "export * as icons from './icons';": '',
              delimiters: ['', ''],
              preventAssignment: false,
            }),
          ] : []
        ),
        ...plugins(pkg, minify)
      ],
      external: ['preact', 'prop-types'],
      output: {
        name: packageName,
        ...(preserveModules
          ? {
              dir: `${outputDir}/${format}`,
            }
          : {
              file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
            }),
        preserveModules,
        format,
        sourcemap: true,
        globals: {
          preact: 'preact',
          'prop-types': 'PropTypes',
        },
      },
    })),
  )
  .flat();

export default configs;
