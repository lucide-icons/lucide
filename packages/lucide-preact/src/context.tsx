import { createContext, type ComponentChildren } from 'preact';
import { useContext } from 'preact/hooks';

const LucideContext = createContext<{
  size?: number;
  fill?: string;
  color?: string;
  strokeWidth?: number;
  absoluteStrokeWidth?: boolean;
}>({
  size: 24,
  fill: 'none',
  color: 'currentColor',
  strokeWidth: 2,
  absoluteStrokeWidth: false,
});

interface LucideProviderProps {
  children: ComponentChildren;
  size?: number;
  fill?: string;
  color?: string;
  strokeWidth?: number;
  absoluteStrokeWidth?: boolean;
}

export function LucideProvider({ children, ...props }: LucideProviderProps) {
  return <LucideContext.Provider value={props}>{children}</LucideContext.Provider>;
}

export const useLucideContext = () => useContext(LucideContext);
