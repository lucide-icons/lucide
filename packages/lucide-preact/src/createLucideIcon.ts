import { ComponentType, FunctionComponent, h, JSX, RefObject, toChildArray } from 'preact';
import defaultAttributes from './defaultAttributes';

type IconNode = [elementName: keyof JSX.IntrinsicElements, attrs: Record<string, string>][]

interface LucideProps extends Partial<Omit<JSX.SVGAttributes, "ref" | "size">> {
  color?: string
  size?: string | number
  strokeWidth?: string | number
}

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string) => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

const createLucideIcon = (iconName: string, iconNode: IconNode): FunctionComponent<LucideProps> => {
  const Component = (
    { color = 'currentColor', size = 24, strokeWidth = 2, children, ...rest }: LucideProps
  ) =>
    h(
      'svg' as unknown as ComponentType<Partial<JSX.SVGAttributes<SVGElement> & { 'stroke-width': number | string }>>,
      {
        ...defaultAttributes,
        width:  String(size),
        height: size,
        stroke: color,
        ['stroke-width' as 'strokeWidth']: strokeWidth,
        class: `lucide lucide-${toKebabCase(iconName)}`,
        ...rest,
      },
      [...iconNode.map(([tag, attrs]) => h(tag, attrs)), ...toChildArray(children)],
    );

  Component.displayName = `${iconName}`;

  return Component;
};

export default createLucideIcon
