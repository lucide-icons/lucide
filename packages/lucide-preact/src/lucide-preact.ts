export * from './icons';
export * as icons from './icons';
export * from './aliases';
export * from './types';

export { default as createLucideIcon } from './createLucideIcon';
export { default as Icon } from './Icon';
