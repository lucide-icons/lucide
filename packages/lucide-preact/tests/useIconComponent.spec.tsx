import { describe, it, expect } from 'vitest';
import { useIconComponent } from '../src/lucide-preact';
import { airVent, coffee } from './testIconNodes';
import { render } from '@testing-library/preact';

describe('Using iconNode component generator', () => {
  it('should create a component from an iconNode', () => {
    const { AirVent } = useIconComponent({ airVent });

    const { container } = render(<AirVent />);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toBeDefined();
  });

  it('should create multiple components', () => {
    const { AirVent, Coffee } = useIconComponent({ airVent, coffee });

    const { container } = render(
      <>
        <AirVent />
        <Coffee />
      </>,
    );

    expect(container).toMatchSnapshot();
    expect(container).toBeDefined();
  });
});
