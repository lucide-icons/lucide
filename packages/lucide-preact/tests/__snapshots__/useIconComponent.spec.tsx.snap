// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using iconNode component generator > should create a component from an iconNode 1`] = `
<svg
  class="lucide lucide-air-vent "
  fill="none"
  height="24"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2"
  />
  <path
    d="M6 8h12"
  />
  <path
    d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12"
  />
  <path
    d="M6.6 15.6A2 2 0 1 0 10 17v-5"
  />
</svg>
`;

exports[`Using iconNode component generator > should create multiple components 1`] = `
<div>
  <svg
    class="lucide lucide-air-vent "
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2"
    />
    <path
      d="M6 8h12"
    />
    <path
      d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12"
    />
    <path
      d="M6.6 15.6A2 2 0 1 0 10 17v-5"
    />
  </svg>
  <svg
    class="lucide lucide-coffee "
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17 8h1a4 4 0 1 1 0 8h-1"
    />
    <path
      d="M3 8h14v9a4 4 0 0 1-4 4H7a4 4 0 0 1-4-4Z"
    />
    <line
      x1="6"
      x2="6"
      y1="2"
      y2="4"
    />
    <line
      x1="10"
      x2="10"
      y1="2"
      y2="4"
    />
    <line
      x1="14"
      x2="14"
      y1="2"
      y2="4"
    />
  </svg>
</div>
`;
